flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveDynamicTableFactory.java

```diff
@@ -70,13 +72,18 @@ public DynamicTableSink createDynamicTableSink(Context context) {
 
         // we don't support temporary hive tables yet
         if (!isHiveTable || context.isTemporary()) {
-            return FactoryUtil.createDynamicTableSink(
-                    null,
-                    context.getObjectIdentifier(),
-                    context.getCatalogTable(),
-                    context.getConfiguration(),
-                    context.getClassLoader(),
-                    context.isTemporary());
+            DynamicTableSink sink =
+                    FactoryUtil.createDynamicTableSink(
+                            null,
+                            context.getObjectIdentifier(),
+                            context.getCatalogTable(),
+                            context.getConfiguration(),
+                            context.getClassLoader(),
+                            context.isTemporary());
+            if (sink instanceof RequireCatalogLock) {
+                ((RequireCatalogLock) sink).setLockFactory(HiveCatalogLock.createFactory(hiveConf));
```

Comment:
For now we have to check and call `RequireCatalogLock#setLockFactory` manually in every `DynamicTableFactory` if necessary. Shall we lift this to `Planner`?

Reply:
Now this is an internal interface. If we need lock to be more general, we can put lock factory into `{@link DynamicTableFactory.Context}`.